src/edu/psu/sweng500/team8/gui/BoardGUI.java

```diff
@@ -0,0 +1,261 @@
+//		gridBagConstraints.insets = new java.awt.Insets(0, 0, 0, 0);
+		add(blocks[0][0], gridBagConstraints);
+
+		blocks[0][1] = new BlockGUI();
+		gridBagConstraints = new java.awt.GridBagConstraints();
+		gridBagConstraints.gridx = 1;
+		gridBagConstraints.gridy = 0;
+		gridBagConstraints.anchor = java.awt.GridBagConstraints.NORTHWEST;
+//		gridBagConstraints.insets = new java.awt.Insets(0, 0, 0, 0);
+		add(blocks[0][1], gridBagConstraints);
+
+		blocks[0][2] = new BlockGUI();
+		gridBagConstraints = new java.awt.GridBagConstraints();
+		gridBagConstraints.gridx = 2;
+		gridBagConstraints.gridy = 0;
+		gridBagConstraints.anchor = java.awt.GridBagConstraints.NORTHWEST;
+//		gridBagConstraints.insets = new java.awt.Insets(0, 0, 0, 0);
+		add(blocks[0][2], gridBagConstraints);
+
+		blocks[1][0] = new BlockGUI();
+		gridBagConstraints = new java.awt.GridBagConstraints();
+		gridBagConstraints.gridx = 0;
+		gridBagConstraints.gridy = 1;
+		gridBagConstraints.anchor = java.awt.GridBagConstraints.NORTHWEST;
+//		gridBagConstraints.insets = new java.awt.Insets(0, 1, 0, 0);
+		add(blocks[1][0], gridBagConstraints);
+
+		blocks[1][1] = new BlockGUI();
+		gridBagConstraints = new java.awt.GridBagConstraints();
+		gridBagConstraints.gridx = 1;
+		gridBagConstraints.gridy = 1;
+		gridBagConstraints.anchor = java.awt.GridBagConstraints.NORTHWEST;
+		add(blocks[1][1], gridBagConstraints);
+
+		blocks[1][2] = new BlockGUI();
+		gridBagConstraints = new java.awt.GridBagConstraints();
+		gridBagConstraints.gridx = 2;
+		gridBagConstraints.gridy = 1;
+		gridBagConstraints.anchor = java.awt.GridBagConstraints.NORTHWEST;
+//		gridBagConstraints.insets = new java.awt.Insets(0, 0, 0, 1);
+		add(blocks[1][2], gridBagConstraints);
+
+		blocks[2][0] = new BlockGUI();
+		gridBagConstraints = new java.awt.GridBagConstraints();
+		gridBagConstraints.gridx = 0;
+		gridBagConstraints.gridy = 2;
+		gridBagConstraints.anchor = java.awt.GridBagConstraints.NORTHWEST;
+//		gridBagConstraints.insets = new java.awt.Insets(0, 1, 1, 0);
+		add(blocks[2][0], gridBagConstraints);
+
+		blocks[2][1] = new BlockGUI();
+		gridBagConstraints = new java.awt.GridBagConstraints();
+		gridBagConstraints.gridx = 1;
+		gridBagConstraints.gridy = 2;
+		gridBagConstraints.anchor = java.awt.GridBagConstraints.NORTHWEST;
+//		gridBagConstraints.insets = new java.awt.Insets(0, 0, 1, 0);
+		add(blocks[2][1], gridBagConstraints);
+
+		blocks[2][2] = new BlockGUI();
+		gridBagConstraints = new java.awt.GridBagConstraints();
+		gridBagConstraints.gridx = 2;
+		gridBagConstraints.gridy = 2;
+		gridBagConstraints.anchor = java.awt.GridBagConstraints.NORTHWEST;
+//		gridBagConstraints.insets = new java.awt.Insets(0, 0, 1, 1);
+		add(blocks[2][2], gridBagConstraints);
+	}
+
+	public void populatePanel(GameSession gameSession, boolean isRefresh) {
+
+		this.gameSession = gameSession;
+		this.selectedCell = null;
+
+		blocks[0][0].populate(gameSession.getGameBoard().getBlock(0),

```

Comment:
- Your argument about OO is great. But those lines were not coded relying
on JIT's optimization. I simply mentioned about JIT because you mentioned
about the compiler/VM's optimization on your earlier email. JIT will
optimize simple calls, but may not optimize complex calls. In terms of
stack push/pop, the argument is never-ending as the stack size and delay
from call vs. memory allocation of the variable. Most of the times, you
don't have to worry about it or code around it in violation of clean code,
unless we are building a system that those things matter.

- The argument about the development cost isn't just about our project, but
I was talking more about in terms of general software development practice.
When you manage a project, you don't want changes to be introduced after
it's coded/tested while there is no measurable merit increase. This is
because the changes, no matter how small, require
re-test/re-evaluation/re-review. A change that can be done in 5 seconds
could bring disastrous consequences. Of course, for our project we can do
whatever we want because it's a sand-box project. But I thought I would
share what I usually do with my projects in terms of Change Management.
What I was going for is 'assuming' calling getGameBoard() multiple times is
a poor design, it isn't worth the change in general software management
perspective. Coding it right in the first place is ideal, but this was
about something else. We can go 'there,' wherever 'there' is. If it's your
way of saying if I coded right in the first we won't have this discussion,
this is my way of saying it's not coded wrong.

- I'm confused about your argument about coupling. It's also most
intriguing one because I've never thought about it that way. As far as I
know, the coupling isn't going to be loosened by calling once and assign it
to a variable. The coupling is already there by having getGameBoard() the
first time. In following example, ClassA has a CBO (coupling between
objects) of 1 with ClassB, and ClassC also has CBO of 1 with Class B. Or at
least that's how tools I've used measured and how I know it should be
measured. If you have a reference showing that multiple calls of the same
method in the same method/class/package should be counted differently in
terms of coupling metrics, please let me know.
//==
Class A {
    /* many times */
    b.getSomethig();
    b.getSomethig();
    ...
    b.getSomethig();
 }

Class C {
    /* just once */
    b.getSomethig();
}

Class B {
   getSomethig();
}
//==

- Moving or changing name is rather a simple change with low risk. You can
just organize the imports and do find/replace in one shot (if done
carefully). However, the assigning to variable approach has a potential
risk in case the implementation of the method being called changes and
return different values and "bypassing them would skip that logic inside of
the method." [5]. Most likely this isn't going to be the case for us. But
we are talking about best practices. And one of the most popular arguments
of 'assign to variable' approach is for the cases where the called method
has long/complex logic (which you brought up, too) but return the same
value isn't going to be the case for us either.

- There is a bunch of stackoverflow pages and others pages discussing this
very topic. Some pros and some againsts. If you read the linked articles
and other articles or talk to your peers, you know it's circumstantial. But
you made it sound like it is a best practice (or has to be done for our
case), so the long email. If you want, we can ask Dr. Barb and ask for his
input as well. In conclusion, it's still unnecessary change(s) for our
project.


[1] [
http://stackoverflow.com/questions/2821140/calling-a-getter-multiple-times-or-calling-once-and-assigning-to-a-variable
]
[2] [
http://stackoverflow.com/questions/2753381/performance-of-vectorsize-is-it-as-fast-as-reading-a-variable
]
[3] [
http://stackoverflow.com/questions/19445737/calling-getters-on-an-object-vs-storing-it-as-a-local-variable-memory-footprin
]
[4] [
http://stackoverflow.com/questions/1923795/java-method-invocation-vs-using-a-variable
]
[5] [
http://programmers.stackexchange.com/questions/181567/should-the-methods-of-a-class-call-its-own-getters-and-setters
]
[6] [https://cs.gmu.edu/~offutt/rsrch/papers/coupmaint.pdf]



Thanks,



Scott Lee



On Mon, Jul 13, 2015 at 8:43 AM, Jeremy Nelson <notifications@github.com>
wrote:

> In src/edu/psu/sweng500/team8/gui/BoardGUI.java
> <https://github.com/garnryang/Team8/pull/13#discussion_r34458616>:
>
> > +
> > +		blocks[2][2] = new BlockGUI();
> > +		gridBagConstraints = new java.awt.GridBagConstraints();
> > +		gridBagConstraints.gridx = 2;
> > +		gridBagConstraints.gridy = 2;
> > +		gridBagConstraints.anchor = java.awt.GridBagConstraints.NORTHWEST;
> > +//		gridBagConstraints.insets = new java.awt.Insets(0, 0, 1, 1);
> > +		add(blocks[2][2], gridBagConstraints);
> > +	}
> > +
> > +	public void populatePanel(GameSession gameSession, boolean isRefresh) {
> > +
> > +		this.gameSession = gameSession;
> > +		this.selectedCell = null;
> > +
> > +		blocks[0][0].populate(gameSession.getGameBoard().getBlock(0),
>
> Agree to disagree I guess. But in the interest of sharing experiences and
> ideas...
>
> I agree premature optimization is generally not a good idea. However I
> would still argue that this is not premature optimization but rather
> PREVENTATIVE optimization.
>
> Object oriented programming is all about abstraction and hiding
> implementation details. Thus, the implementor of the method should not need
> to know all the places their method is used, and the caller should not need
> to know how the method is implemented. Suppose then that the implementor
> changes getGameBoard() to do something costly. Now all the assumptions
> about JIT optimization go out the window. The implementor might make an
> assumption that the method is not called very often, so it is not a
> concern. However, this introduces a huge performance slowdown because it is
> being called many times unnecessarily.
>
> I don't think your argument about time spent making this change holds any
> water. It would take ~5 seconds to make the change. If you write code like
> this in the first place, it would require less keyboard characters and
> theoretically less time. Let's not go there.
>
> As for being in the middle of the development cycle -- this means methods
> will move and change and be renamed frequently, yes? Each one of these
> calls to getGameBoard is an instance of coupling between the classes. As we
> learned in 537/585/588, coupling needs to be reduced to make changes
> easier. In this case, if the method was moved or renamed, you would have to
> change it in 9 places rather than 1, costing more time.
>
> As for the variable getting assigned to, the assumption this code is
> making is that the gameBoard is the same for each call. Without that, it
> would not make sense. Thus to ensure it is using the same gameBoard, it
> should only query it once.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/garnryang/Team8/pull/13/files#r34458616>.
>